Documentation/design-docs/standalone-gc-loading.md

```diff
@@ -0,0 +1,222 @@
+that the runtime have a protocol for locating and loading such GCs. The JIT is capable of being loaded
+in this manner and, because of this, a significant amount of prior art exists for loading components
+for shared libraries from the file system. This specification is based heavily on the ways that a
+standalone JIT can be loaded.
+
+Fundamentally, the algorithm for loading a standalone GC consists of these steps:
+
+0. Identify whether or not we should be using a standalone GC at all.
+1. Identify *where* we should be looking for GC shared libraries.
+2. Identify *what* we are looking for - the specific name of a dynamic shared library that
+   will be loaded.
+3. Load the dynamic shared library and ask it to identify itself (name and version).
+4. Check that the version numbers are compatible.
+5. If so, initialize the GC and continue on with EE startup. If not, reject the dynamic shared library
+   and raise an appropriate user-visible error.
+
+The algorithm for initializing the DAC against a target process using a standalone GC consists of these steps:
+
+1. Identify whether or not the target process is using a standalone GC at all. If not, no further
+   checks are necessary.
+2. If so, inspect the version number of the standalone GC in the target process and determine whether
+   or not the DAC is compatible with that version. If not, present a notification of some kind
+   that the debugging experience will be degraded.
+3. Continue onwards.
+
+Each one of these steps will be explained in detail below.
+
+### Identifying candidate shared libraries
+
+The question of whether or not the EE should attempt to locate and load a standalone GC
+is answered by the EE's configuration system (`EEConfig`). EEConfig has the ability to
+query configuration information from a variety of sources, such a registry keys or
+environment variables. Using this subsystem, users can specify a specific environment
+variable to indicate that they are interested in loading a standalone GC.
+
+There are three configuration keys that control the behavior of loading a standalone GC:
+
+1. `UseStandaloneGC`, (environment variable `COMPLUS_UseStandaloneGC`). If set and equal to "1", the EE
+   will understand that it is to load a standalone GC from somewhere and begin to search for the GC to
+   load.
+2. `StandaloneGCLocation`, (environment variable `COMPlus_StandaloneGCLocation`). If set, it will influence
+   the path that the EE will use to probe for a standalone GC dynamic shared library.
+3. `StandaloneGCName`, (environment variable `COMPlus_StandaloneGCName`). If set, it will influence the name
+   of the dynamic shared library that the EE will look for.
+
+It is not an error to emit either `StandaloneGCLocation` or `StandaloneGCName`. If not specified, and
+`UseStandaloneGC` is set, `StandaloneGCLocation` will default to the same directory as CoreCLR and `StandaloneGCName`
+will default to `gc` (`gc.dll` on windows, `libgc.so/dylib` on Linux/OSX).
```

Comment:
Perhaps the default library name should be `clrgc`, not just `gc`?

Reply:
Good point, especially to avoid conflicts with the BDW GC.

Reply:
I think I'm going to remove this part from the design doc, but we definitely should be using `clrgc` as the default name in the implementation.